core/spec/lib/spree/core/controller_helpers/payment_parameters_spec.rb

```diff
@@ -77,4 +78,103 @@ class FakeController < ApplicationController
       it('returns the original hash') { expect(subject).to eq(params) }
     end
   end
+
+  describe '#move_existing_card_into_payments_attributes' do
+    subject do
+      controller.move_existing_card_into_payments_attributes(params)
+    end
+
+    let(:params) do
+      ActionController::Parameters.new(
+        order: {
+          existing_card: '123',
+          other_order_param: 1,
+        },
+        cvc_confirm: '456',
+        other_param: 2,
+      )
+    end
+
+    it 'produces the expected hash' do
+      expect(subject).to eq(
+        ActionController::Parameters.new(
+          order: {
+            payments_attributes: [
+              {
+                source_attributes: {
+                  existing_card_id: '123',
+                  verification_value: '456'
+                },
+              },
+            ],
+            other_order_param: 1,
+          },
+          other_param: 2,
+        )
+      )
+    end
+
+    context 'when cvc_confirm is missing' do
+      before { params.delete(:cvc_confirm) }
+
+      it 'produces the expected hash' do
+        expect(subject).to eq(
+          ActionController::Parameters.new(
+            order: {
+              payments_attributes: [
+                {
+                  source_attributes: {
+                    existing_card_id: '123',
+                    verification_value: nil,
+                  },
+                },
+              ],
+              other_order_param: 1,
+            },
+            other_param: 2,
+          )
+        )
+      end
+    end
+
+    context 'when order params are missing' do
+      before { params.delete(:order) }
+      it('returns the original hash') { expect(subject).to eq(params) }
+    end
+
+    context 'when existing_card is missing' do
+      before { params[:order].delete(:existing_card) }
+      it('returns the original hash') { expect(subject).to eq(params) }
+    end
+
+  end
+
+  describe '#set_payment_parameters_amount' do
+    subject do
+      controller.set_payment_parameters_amount(params, order)
+    end
+
+    let(:params) do
+      ActionController::Parameters.new(
+        order: {
+          payments_attributes: [{}],
+          other_order_param: 1,
+        },
+        other_param: 2,
+      )
+    end
+    let(:order) { create(:order_with_line_items, line_items_price: 101.00, line_items_count: 1, shipment_cost: 0) }
```

Comment:
Sorry to be picky on lots of good work. 

This factory is very expensive, and I do not see why persisted data would be needed at all in a controller helper class. Maybe use `build_stubbed` or Order.new if possible.